File: src/test/java/com/runninglane/dto/buddy/test/java/cases/generic/multiple/MultipleTypeParametersTest.java
package com.runninglane.dto.buddy.test.java.cases.generic.multiple;

import com.runninglane.dto.buddy.javainterop.DtoBuddy;
import com.runninglane.dto.buddy.javainterop.bytecode.ThreeStepsByteCodeStrategy;
import com.runninglane.dto.buddy.javainterop.bytecode.compile.CompileJavaByteCodeStrategyCompliment;
import com.runninglane.dto.buddy.test.java.cases.generic.GenericTypeTestHelper;
import org.junit.jupiter.api.Test;

import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import static org.junit.jupiter.api.Assertions.*;

public class MultipleTypeParametersTest {
    private final DtoBuddy dtoBuddy = new DtoBuddy(
        new ThreeStepsByteCodeStrategy<>(
            new CompileJavaByteCodeStrategyCompliment()
        )
    );

    private final GenericTypeTestHelper helper = new GenericTypeTestHelper();

    @Test
    public void testImplement() {
        Class<?> concreteClass = dtoBuddy.implement(
            DtoInterfaceWithMultipleTypeParameters.class,
            List.of(String.class, Integer.class, String.class, Long.class)
        );

        assertNotNull(concreteClass);
        assertTrue(DtoInterfaceWithMultipleTypeParameters.class.isAssignableFrom(concreteClass));
        assertEquals(String.class, helper.findMostSpecificMethodByName(concreteClass, "getSimple").getReturnType());
        assertEquals(List.class, helper.findMostSpecificMethodByName(concreteClass, "getList").getReturnType());
        assertEquals(Map.class, helper.findMostSpecificMethodByName(concreteClass, "getMap").getReturnType());
    }

    @Test
    public void testCreate() {
        // This test will fail until create() is implemented
        Map<String, Object> params = new HashMap<>();
        params.put("simple", "Simple String");
        params.put("list", List.of(1, 5, 9));
        Map<String, Long> mapParam = new HashMap<>();
        mapParam.put("key1", 1L);
        mapParam.put("key2", 2L);
        mapParam.put("key3", 3L);
        params.put("map", mapParam);

        Class<?> concreteClass = dtoBuddy.implement(
            DtoInterfaceWithMultipleTypeParameters.class,
            List.of(String.class, Integer.class, String.class, Long.class)
        );
        DtoInterfaceWithMultipleTypeParameters<String, Integer, String, Long> dto = dtoBuddy.create(concreteClass, params);

        assertNotNull(dto);
        assertEquals("Simple String", dto.getSimple());
        assertEquals(List.of(1, 5, 9), dto.getList());
        assertEquals(mapParam, dto.getMap());
    }

    @Test
    public void testCreateThenPopulate() {
        // This test will fail until populate() is implemented
        Map<String, Object> params = new HashMap<>();
        Map<String, Long> mapParam = new HashMap<>();
        mapParam.put("key4", 4L);
        mapParam.put("key5", 5L);
        params.put("map", mapParam);

        Class<?> concreteClass = dtoBuddy.implement(
            DtoInterfaceWithMultipleTypeParameters.class,
            List.of(String.class, Integer.class, String.class, Long.class)
        );

        Map<String, Object> createParams = new HashMap<>();
        createParams.put("simple", "Initial String");
        createParams.put("list", List.of(1, 2, 3));
        createParams.put("map", null);

        DtoInterfaceWithMultipleTypeParameters<String, Integer, String, Long> dto = dtoBuddy.create(concreteClass, createParams);

        assertNotNull(dto);
        dtoBuddy.populate(dto, params);
        assertEquals("Initial String", dto.getSimple());
        assertEquals(List.of(1, 2, 3), dto.getList());
        assertEquals(mapParam, dto.getMap());
    }

        @Test
    public void testCreateWithNullValues() {
            // This test will fail until create() is implemented
            Map<String, Object> params = new HashMap<>();
            params.put("simple", "Test String");
            params.put("list", List.of(10, 20, 30));
            params.put("map", null);

            Class<?> concreteClass = dtoBuddy.implement(
                DtoInterfaceWithMultipleTypeParameters.class,
                List.of(String.class, Integer.class, String.class, Long.class)
            );

            DtoInterfaceWithMultipleTypeParameters<String, Integer, String, Long> dto = dtoBuddy.create(concreteClass, params);

            assertNotNull(dto);
            assertEquals("Test String", dto.getSimple());
            assertEquals(List.of(10, 20, 30), dto.getList());
            assertNull(dto.getMap());
        }

        @Test
        public void testCreateThenPopulateWithMultipleValues() {
            Map<String, Object> params = new HashMap<>();
            params.put("simple", "Updated String");
            params.put("list", List.of(100, 200, 300));
            Map<String, Long> mapParam = new HashMap<>();
            mapParam.put("keyA", 10L);
            mapParam.put("keyB", 20L);
            params.put("map", mapParam);

            Class<?> concreteClass = dtoBuddy.implement(
                DtoInterfaceWithMultipleTypeParameters.class,
                List.of(String.class, Integer.class, String.class, Long.class)
            );

            Map<String, Object> createParams = new HashMap<>();
            createParams.put("simple", "Original String");
            createParams.put("list", List.of(1));
            Map<String, Long> initialMap = new HashMap<>();
            initialMap.put("key1", 1L);
            createParams.put("map", initialMap);

            DtoInterfaceWithMultipleTypeParameters<String, Integer, String, Long> dto = dtoBuddy.create(concreteClass, createParams);

            assertNotNull(dto);

            dtoBuddy.populate(dto, params);

            assertEquals("Updated String", dto.getSimple());
            assertEquals(List.of(100, 200, 300), dto.getList());
            assertEquals(mapParam, dto.getMap());
        }
    }

File: src/test/kotlin/com/runninglane/dto/buddy/test/cases/generic/multiple/MultipleTypeParametersTest.kt
package com.runninglane.dto.buddy.test.cases.generic.multiple

import com.runninglane.dto.buddy.DtoBuddy
import com.runninglane.dto.buddy.test.NamingStrategyWithCountUpSuffix
import kotlin.reflect.full.isSubclassOf
import kotlin.reflect.full.memberProperties
import kotlin.reflect.jvm.jvmErasure
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertNotNull
import kotlin.test.assertTrue

class MultipleTypeParametersTest {
    private val dtoBuddy = DtoBuddy(NamingStrategyWithCountUpSuffix())

    @Test
    fun testImplement() {
        run {
            val concreteClass = dtoBuddy.implement(
                DtoInterfaceWithMultipleTypeParameters::class,
                listOf(String::class, Int::class, String::class, Long::class)
            )

            assertNotNull(concreteClass)
            assertTrue(concreteClass.isSubclassOf(DtoInterfaceWithMultipleTypeParameters::class))
            assertEquals(String::class, concreteClass.memberProperties.single { it.name == "simple" }.returnType.jvmErasure)
            assertEquals(List::class, concreteClass.memberProperties.single { it.name == "list" }.returnType.jvmErasure)
            assertEquals(Map::class, concreteClass.memberProperties.single { it.name == "map" }.returnType.jvmErasure)
        }
    }

    @Test
    fun testCreate() {
        // This test will fail until create() is implemented
        val params = mapOf(
            "simple" to "Simple String",
            "list" to listOf(1, 5, 9),
            "map" to mapOf("key1" to 1L, "key2" to 2L, "key3" to 3L)
        )

        val concreteClass = dtoBuddy.implement(
            DtoInterfaceWithMultipleTypeParameters::class,
            typeParams = listOf(String::class, Int::class, String::class, Long::class)
        )
        val dto: DtoInterfaceWithMultipleTypeParameters<String, Int, String, Long> = dtoBuddy.create(concreteClass, params)

        assertNotNull(dto)
        assertEquals("Simple String", dto.simple)
        assertEquals(listOf(1, 5, 9), dto.list)
        assertEquals(mapOf("key1" to 1L, "key2" to 2L, "key3" to 3L), dto.map)
    }

    @Test
    fun testCreateThenPopulate() {
        // This test will fail until populate() is implemented
        val params = mapOf(
            "map" to mapOf("key4" to 4L, "key5" to 5L)
        )

        val concreteClass = dtoBuddy.implement(
            DtoInterfaceWithMultipleTypeParameters::class,
            typeParams = listOf(String::class, Int::class, String::class, Long::class)
        )
        val dto: DtoInterfaceWithMultipleTypeParameters<String, Int, String, Long> = dtoBuddy.create(
            concreteClass,
            mapOf("simple" to "Initial String", "list" to listOf(1, 2, 3), "map" to null)
        )

        assertNotNull(dto)
        dtoBuddy.populate(dto, params)
        assertEquals("Initial String", dto.simple)
        assertEquals(listOf(1, 2, 3), dto.list)
        assertEquals(mapOf("key4" to 4L, "key5" to 5L), dto.map)
    }

    @Test
    fun testCreateWithNullValues() {
        // This test will fail until create() is implemented
        val params = mapOf(
            "simple" to "Test String",
            "list" to listOf(10, 20, 30),
            "map" to null
        )

        val concreteClass = dtoBuddy.implement(
            DtoInterfaceWithMultipleTypeParameters::class,
            typeParams = listOf(String::class, Int::class, String::class, Long::class)
        )
        val dto: DtoInterfaceWithMultipleTypeParameters<String, Int, String, Long> = dtoBuddy.create(concreteClass, params)

        assertNotNull(dto)
        assertEquals("Test String", dto.simple)
        assertEquals(listOf(10, 20, 30), dto.list)
        assertEquals(null, dto.map)
    }

    @Test
    fun testCreateThenPopulateWithMultipleValues() {
        // This test will fail until populate() is implemented
        val params = mapOf(
            "simple" to "Updated String",
            "list" to listOf(100, 200, 300),
            "map" to mapOf("keyA" to 10L, "keyB" to 20L)
        )

        val concreteClass = dtoBuddy.implement(
            DtoInterfaceWithMultipleTypeParameters::class,
            typeParams = listOf(String::class, Int::class, String::class, Long::class)
        )
        val dto: DtoInterfaceWithMultipleTypeParameters<String, Int, String, Long> = dtoBuddy.create(
            concreteClass,
            mapOf("simple" to "Original String", "list" to listOf(1), "map" to mapOf("key1" to 1L))
        )

        assertNotNull(dto)

        dtoBuddy.populate(dto, params)

        assertEquals("Updated String", dto.simple)
        assertEquals(listOf(100, 200, 300), dto.list)
        assertEquals(mapOf("keyA" to 10L, "keyB" to 20L), dto.map)
    }
}